### DISCUSSION TITLE
Dynamic display output configuration

### DISCUSSION THREAD

**Discussion by RangerMauve** [YaLTeR/niri#327] [Ideas]
I use a portable computer with a built in screen and when I plug in my dock or external monitor I'd like to have the built in monitor go off.

Some stuff I'm unsure of:

When the dock gets disconnected the windows on it seem to be "lost" but still in memory and show back up only when I re-dock. Feels like it'd be good to have orphaned windows moved to the active workspace on the first remaining monitor.

It feels like there could be an opportunity for setting conditional output configs like "if both these displays exist, here's the per output rules".

In the meantime I'll see if I can set up script to modify the config dynamically when detecting a new monitor.

**Comment by YaLTeR** [comment]
Hey, yeah, transient output config via IPC (and via wlr output management protocol) is planned, and also automatic disable of laptop panel on lid close is planned too #145.

**Reply by YaLTeR** [reply]
Transient output config is now implemented in 65b9c74f624b4a78c2247135bf1a96dcac0bd009. `niri msg output eDP-1 off` for example

**Reply by RangerMauve** [reply]
That's great t hear! Are you looking for PRs for stuff like this? I'm not great at rust but I'm hoping to settle on niri as my main window manager so it's more likely I'll have need to tinker with it.

**Reply by YaLTeR** [reply]
You could try the automatic laptop lid disable. Libinput should expose a lid switch event for this in smithay

**Reply by RangerMauve** [reply]
I sadly don't have a computer with a lid (I main GPD Win 4) but the event system there seems like a good lead in general, ty

**Comment by RangerMauve** [comment]
Gonna attempt to use this: https://sr.ht/~emersion/kanshi/

**Reply by RangerMauve** [reply]
Dang. Getting this error:

```
compositor doesn't support wlr-output-management-unstable-v1
```

**Comment by tv42** [comment]
> When the dock gets disconnected the windows on it seem to be "lost" but still in memory and show back up only when I re-dock.

When I disconnect by USB-C or HDMI monitor connections, I think[1] what happens is that the workspaces on those monitors get added to the end of the list of workspaces on my laptop built-in display, with all the windows there.

[1]: Haven't repeated the experiment to say this authoritatively, but I don't think I "lost" any windows.

**Reply by YaLTeR** [reply]
When you disconnect a monitor, all workspaces move to the end of one of the existing monitors, then upon reconnecting they will move back to the original output.

**Comment by druskus20** [comment]
It would be nice to have a way to configure output settings dynamically.

I have a laptop that I use in different ways:
- built-in output
- built-in + 1 output
- built-in + 2 outputs
- built-in + 2 outputs (flipped)


It would be nice to have control over things such as: 
- dynamically assigning workspaces to outputs
- output position, scale, resolution...

My current approach is to comment/uncomment some lines in the config whenever I switch workplace.

**Reply by YaLTeR** [reply]
> output position, scale, resolution...

You can use kanshi for this. niri implements the necessary protocols since some time ago.

**Reply by RangerMauve** [reply]
Just tried setting up kanshi again and it's working great!

**Reply by druskus20** [reply]
Kanshi is working super well, I did not know about it.
One more thing I'd love is the ability to move all "things" (windows / workspaces / columns maybe?) from one output to another. 
I think maybe something can be made with `niri msg action`, but I'm not exactly sure how that looks.

**Reply by YaLTeR** [reply]
Should be possible to get the list of workspaces and windows from niri msg windows and niri msg workspaces, then use the action IPC to move them like niri msg action move-window-to-output --id 5 DP-2 (workspaces by id addressing exists only in raw json IPC rn)

**Reply by ornanovitch** [reply]
> I have a laptop that I use in different ways:
> 
>     built-in output
>     built-in + 1 output
>     built-in + 2 outputs
>     built-in + 2 outputs (flipped)
> 
> It would be nice to have control over things such as:
> 
>     dynamically assigning workspaces to outputs

I have the same kind of setup, I switched to Niri and the only thing I miss is the ability to assign workspace to multiple outputs with a mechanism of priority order

```kdl
workspace "stuff" { // will open on DP-2 if DP-1 is off
    open-on-output "DP-1"
    open-on-output "DP-2"
}
